Bio/Seq.py

```diff
@@ -1056,220 +1148,326 @@ def rstrip(self, chars=None):
             raise TypeError(
                 "argument must be None or a string, Seq, MutableSeq, or bytes-like object"
             ) from None
-        return Seq(data)
+        if inplace:
+            if not isinstance(self._data, bytearray):
+                raise TypeError("Sequence is immutable")
+            self._data[:] = data
+            return self
+        elif isinstance(self, UnknownSeq):
+            return Seq(data)
+        else:
+            return self.__class__(data)
 
-    def upper(self):
-        """Return an upper case copy of the sequence.
+    def upper(self, inplace=False):
```

Comment:
Why do you have the inplace argument on the ABC? This is specific to the MutableSeq only isn't it?

Reply:
This is to be able to give a clear error message. If the inplace argument is on the MutableSeq only, and somebody inadvertently uses inplace=True on a Seq object, then it will raise TypeError saying that inplace is an unexpected keyword argument. With inplace on the ABC,  the error message explains that inplace=True cannot be used on a Seq object because it is immutable. This is clearer in my opinion.

Reply:
Counter argument: The exception being immutable if ``inplace`` is True is only seen at run time. Having the often inappropriate optional argument on the base class will make the argument appear in the documentation and autocompletion in IDEs, and may also confuse dynamic code inspection for error analysis.

I think our preferences on clarity are opposite on this point.

Reply:
I do think it's important for the `Seq` class to handle `inplace=True` arguments gracefully. Otherwise the resulting error message (saying that `inplace` is an unknown argument) is way too confusing, especially if this error occurs somewhere deep in some complex piece of code. The user may think that something is wrong with his Biopython version.

One option is to implement the base class in C, which gives us a lot more flexibility in how to handle arguments (e.g. we could silently accept `inplace=False` with a `Seq` object). Not sure if idoing so is worth the effort just for this application.

Reply:
I would agree with avoiding C for this base class.

Reply:
The other option is to have a `Seq` class only, no `MutableSeq` or base class, and let `Seq` objects handle both immutable and mutable sequences, with an attribute `mutable` showing whether the sequence is mutable or not.

Reply:
That might have been a good starting point, but that kind of change now seems too much from a backward compatibility point of view.

Reply:
I have described three options:

1)  Current PR: raise a `TypeError` if `inplace=True` is used on a `Seq` object.

Counter argument: The exception being immutable if `inplace` is `True` is only seen at run time. Having the often inappropriate optional argument on the base class will make the argument appear in the documentation and autocompletion in IDEs, and may also confuse dynamic code inspection for error analysis.

Can you explain this further? The `inplace` argument would still valid be on `Seq` objects: `Seq("abcd").upper(inplace=False)` simply returns `Seq("ABCD")`, without raising a `TypeError`. Arguments with restrictions on acceptable values are very common in Python. For example, `"ABCD".find(5)` gives `TypeError: must be str, not int`.

2) Implement the base class in C, and give an informative error message if `inplace=True` is used on a `Seq` object.

I agree that this is too invasive at this point.

3) Let `Seq` objects handle both immutable and mutable sequences.

Counter argument: That kind of change now seems too much from a backward compatibility point of view.

I don't quite understand this either. Any existing scripts would treat `Seq` objects as immutable, and nothing changes for `Seq` objects used with immutable sequences. So from a backward compatibility viewpoint, I don't see a problem here.

Reply:
@peterjc  Any updates?